parachain/pallets/score-staking/src/lib.rs

```diff
@@ -445,6 +427,70 @@ pub mod pallet {
 			let payment = Scores::<T>::get(&account).ok_or(Error::<T>::UserNotExist)?;
 			Self::claim(origin, payment.unpaid_reward)
 		}
+
+		#[pallet::call_index(9)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal))]
+		pub fn update_total_staking_amount(
+			origin: OriginFor<T>,
+			account_id: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let _ = T::TEECallOrigin::ensure_origin(origin)?;
+
+			match Scores::<T>::get(&account_id) {
+				Some(mut s) => {
+					s.total_staking_amount = amount;
+					Scores::<T>::insert(account_id.clone(), s);
+					Self::deposit_event(Event::TotalStakingAmountUpdated {
+						account_id: account_id.clone(),
+						amount,
+					});
+					Ok(Pays::No.into())
+				},
+				None => Err(Error::<T>::UserNotExist.into()),
+			}
+		}
+
+		#[pallet::call_index(10)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal))]
+		pub fn distribute_rewards(
+			origin: OriginFor<T>,
+			round_index: RoundIndex,
+		) -> DispatchResultWithPostInfo {
+			let _ = T::TEECallOrigin::ensure_origin(origin)?;
+			ensure!(
+				round_index > LastTokenDistributionRound::<T>::get(),
+				Error::<T>::RoundRewardsAlreadyDistributed
+			);
+			let round_reward: BalanceOf<T> = (T::YearlyInflation::get() * T::YearlyIssuance::get()
+				/ YEARS.into()) * Self::round_config().interval.into();
+			let round_reward_u128 = round_reward.saturated_into::<u128>();
+			let total_stake_u128 = ParaStaking::Pallet::<T>::total().saturated_into::<u128>();
+			let total_score = Self::total_score();
+			let n = Self::round_config().stake_coef_n;
+			let m = Self::round_config().stake_coef_m;
+
+			for (a, mut p) in Scores::<T>::iter() {
```

Comment:
This loop is quite expensive, it's going to be O(n), 
Have we researched using something like `on_idle` hook for this, that way we make sure to use the idle blockspace and that way we allow the chain to be used for other activities and not prioritise the distribution of tokens?

Reply:
I had the same concern in the original PR (the current `on_initialize` hook), iterating over the entries could be a problem eventually. The issue with `on_idle` is that it's only executed if there is weight left, and the logic implemented needs to consider the remaining weight as it should not exceed it. But the main problem with using a hook is that we have no access to the IDGraphs to calculate the total staking for each user.

Another solution could be to go back to calculating the user's rewards one by one (this was the initial proposal of the [PR](https://github.com/litentry/litentry-parachain/pull/3074) that was closed), but in order to have a consistent state when calculating the rewards for each users we also keep track of the `total_score` and `total_staking` of the last round and use that in the calculation. What do you think @Kailai-Wang ?

Reply:
Another approach we could explore is, if we calculate the users reward when the user claims rewards. If we know the total reward of that round, and how much is staked by the user, then we can simply calculate the reward the user has to receive(provided we have storage items for it). This is just an approach, I can't think of exact details. I think this is what polkadot staking or the frame staking pallet does. I simply want to avoid an O(N) loop if possible.

Reply:
With the logic that we have to calculate the rewards we would need to keep track of values like `total_stake`, `total_score` for every user each round as well as the global values we use, so we will end up in a similar situation (or even worst). I think the initial approach should be fine, It's just that back then we didn't think about how to keep the state consistent between rounds to do the calculations, keeping track of the global values ​​would solve that.